presto-main/src/main/java/com/facebook/presto/execution/DataDefinitionExecution.java

```diff
@@ -65,6 +68,21 @@ private DataDefinitionExecution(
         this.accessControl = requireNonNull(accessControl, "accessControl is null");
         this.stateMachine = requireNonNull(stateMachine, "stateMachine is null");
         this.parameters = parameters;
+
+        stateMachine.addStateChangeListener(state -> {
```

Comment:
Can there be some race conditions on SPI calls? For example can `beginQuery` be called after some metadata call? For example after `beginCreateTable` or `getNewTableLayout` or even after first `PageSinks`/`PageSources` being created on the workers?